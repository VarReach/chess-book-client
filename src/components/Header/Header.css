.app__header {
  background-color: #2f343e;
  height: 96px;
  color: #f7f7f7;
  box-shadow: 0 5px 0 0 #d9d9d9;
}

.app__header-container {
  display: flex;
  align-items: center;
}

.app__header-title {
  display: inline-flex;
  align-items: center;
  font-family: 'Libre Baskerville', serif;
  font-size: 30px;
  color: #f7f7f7;
  text-decoration: none;
  letter-spacing: 0.03em;
  padding-right: 20px;
}

.app__header-nav {
  display: inline-block;
}

.app__header-nav {
  margin: 0;
  height: 100%;
  width: 100%;
  display: inline-flex;
  align-items: center;
}

.app__header-nav a, .app__header-nav-user-button {
  padding: 10px 15px;
  text-decoration: none;
  color: #f7f7f7;
}

.app__header-nav>a, .app__header-nav-user-button {
  margin: 0 5px;
  border-radius: 5px;
  font-weight: 600;
  font-size: 15px;
}

.app__header-nav-user {
  position: relative;
  margin-left: auto;
}

.app__header-nav-user-button {
  border: none;
  font-family: 'Montserrat', sans-serif;
  background-color: transparent;
  outline: none;
}

.app__header-nav-user-button:focus {
  background-color: #4d5b75;
}

.app__header-nav-user-button:hover
 {
  cursor: pointer;
}


.app__header-nav>a:hover, .app__header-nav-user-button:hover {
  background-color: #4d5b75;
}

#app__header-user-dropdown {
  width: 140px;
  height: 100px;
  top: 50px;
  right: 0;
}

#app__header-user-dropdown:after {
	bottom: 100%;
	right: 22px;
	border: solid transparent;
	content: " ";
	height: 0;
	width: 0;
	position: absolute;
	pointer-events: none;
	border-color: #00000000;
	border-bottom-color: #242831;
	border-width: 0 16px 8px 16px;
	margin-left: -20px;
}

#app__header-user-dropdown a {
  margin: 0;
  display: inline-block;
  width: 100%;
  font-weight: normal;
  font-size: 12px;
}

#app__header-user-dropdown a:hover, #app__header-user-dropdown a:focus {
  background-color: #363e53;
  outline: none;
} 

/* .app__header-nav-list a:hover {
  border-bottom: 1px solid #f7f7f7;
} */

@media only screen and (max-width: 600px) {
  .app__header-title {
    width: 60px;
    margin: 0 5px;
  }

  .app__header-title>span {
    display:none;
  }
}