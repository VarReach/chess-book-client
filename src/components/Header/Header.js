import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import TokenService from '../../services/token-service';
import { Link } from 'react-router-dom';
import logo from '../../images/ChessBlogLogo.png';
import './Header.css';
import UserContext from '../../contexts/UserContext';

export default class Header extends Component {
  state = {
    showDropDown: false,
  }

  static contextType = UserContext;

  componentWillUnmount() {
    if (this.state.showDropDown) {
      document.removeEventListener('click', this.closeDropDown);
    }
  }

  toggleDropDown = (e) => {
    if (e && ReactDOM.findDOMNode(this.refs.dropdown).contains(e.target)) {
      return;
    }
    this.state.showDropDown ? this.hideDropDown() : this.showDropDown();
  }

  showDropDown = () => {
    this.setState({ showDropDown: true }, () => {document.addEventListener('click', this.hideDropDown)});
  }

  hideDropDown = () => {
    this.setState({ showDropDown: false }, () => {document.removeEventListener('click', this.hideDropDown)});
  }

  render() {
    return (
      <nav role="navigation" className="app__header">
        <div className="container app__header-container">
          <Link to="/" className="app__header-title">
            <img src={logo} alt="Chess Piece: Rook"/>
            <span>ChessBook</span>
          </Link> 
          <div className="app__header-nav">
            {TokenService.hasAuthToken() && TokenService.hasPermissions() && <Link to="/editor">Editor</Link>}
            <div className="app__header-nav-user">
              <button className="app__header-nav-user-button" onClick={this.toggleDropDown}>
                {this.context.user && this.context.user.user_name
                  ? <><i className="fas fa-user"/><span>{this.context.user.user_name}</span></>
                  : <><i className="fas fa-user-slash"/><span>{'Guest'}</span></>}
                <i className="fas fa-caret-down"/>
              </button>
              <div ref="dropdown" className={"dropdown app__header-user-dropdown" + (!this.state.showDropDown ? ' hidden' : '')}>
                <ul>
                 {TokenService.hasAuthToken()
                    ? <li><Link to ="/" onClick={this.context.logout}>Logout</Link></li>
                    : (
                        <>
                          <li><Link to="/login"><i className="fas fa-sign-in-alt"/>Log in</Link></li>
                          <li><Link to="/login"><i className="fas fa-user-plus"/>Sign up</Link></li>
                        </>
                    )}
                </ul>
              </div>
            </div>
          </div>
        </div>
      </nav>
    );
  }
}