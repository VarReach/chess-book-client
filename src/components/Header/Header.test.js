import React from 'react';
import { shallow, mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import toJson from 'enzyme-to-json';
import Header from './Header';

describe('Header component', () => {
  it('renders the complete header', () => {
    const wrapper = shallow(<Header/>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('no editor link when user doesn\'t have permissions', () => {
    const editorLink = shallow(<Header />)
      .find('#app__header-editor-link');
    expect(toJson(editorLink)).toMatchSnapshot();
  });

  it('no dropdown by default', () => {
    const dropDown = shallow(<Header />)
      .find('#app__header-user-dropdown');
    expect(toJson(dropDown)).toMatchSnapshot();
  });

  it('shows dropdown on button click', () => {
    const header = mount(
      <MemoryRouter>
        <Header /> 
      </MemoryRouter>
    );
    header.find('.app__header-nav-user-button').simulate('click');
    expect(toJson(header.find('#app__header-user-dropdown'))).toMatchSnapshot();
    header.unmount();
  });
});