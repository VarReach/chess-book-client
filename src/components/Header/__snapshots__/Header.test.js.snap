// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component no dropdown by default 1`] = `null`;

exports[`Header component no editor link when user doesn't have permissions 1`] = `null`;

exports[`Header component renders the complete header 1`] = `
<nav
  className="app__header"
  role="navigation"
>
  <div
    className="container app__header-container"
  >
    <Link
      className="app__header-title"
      replace={false}
      to="/"
    >
      <img
        alt="Chess Piece: Rook"
        src="ChessBlogLogo.png"
      />
      <span>
        ChessBook
      </span>
    </Link>
    <div
      className="app__header-nav"
    >
      <div
        className="app__header-nav-user"
      >
        <button
          className="app__header-nav-user-button"
          onClick={[Function]}
        >
          <i
            className="fas fa-user-slash"
          />
          <span>
            Guest
          </span>
          <i
            className="fas fa-caret-down"
          />
        </button>
        <DropDown />
      </div>
    </div>
  </div>
</nav>
`;

exports[`Header component shows dropdown on button click 1`] = `
<div
  className="dropdown"
  id="app__header-user-dropdown"
>
  <ul>
    <li>
      <Link
        replace={false}
        to="/login"
      >
        <a
          href="/login"
          onClick={[Function]}
        >
          <i
            className="fas fa-sign-in-alt"
          />
          Log in
        </a>
      </Link>
    </li>
    <li>
      <Link
        replace={false}
        to="/register"
      >
        <a
          href="/register"
          onClick={[Function]}
        >
          <i
            className="fas fa-user-plus"
          />
          Sign up
        </a>
      </Link>
    </li>
  </ul>
</div>
`;
